File: app/src/main/java/br/com/jonathanzanella/myexpenses/bill/BillRepository.java
package br.com.jonathanzanella.myexpenses.bill;

import android.support.annotation.NonNull;
import android.support.annotation.WorkerThread;

import org.apache.commons.lang3.StringUtils;
import org.joda.time.DateTime;

import java.util.List;
import java.util.UUID;

import br.com.jonathanzanella.myexpenses.database.Fields;
import br.com.jonathanzanella.myexpenses.database.ModelRepository;
import br.com.jonathanzanella.myexpenses.database.Repository;
import br.com.jonathanzanella.myexpenses.database.Where;
import br.com.jonathanzanella.myexpenses.expense.Expense;
import br.com.jonathanzanella.myexpenses.expense.ExpenseRepository;
import br.com.jonathanzanella.myexpenses.log.Log;
import br.com.jonathanzanella.myexpenses.validations.ValidationError;
import br.com.jonathanzanella.myexpenses.validations.ValidationResult;

public class BillRepository  implements ModelRepository<Bill> {
	private final Repository<Bill> repository;
	private final BillTable billTable = new BillTable();
	private final ExpenseRepository expenseRepository;

	public BillRepository(Repository<Bill> repository, ExpenseRepository expenseRepository) {
		this.repository = repository;
		this.expenseRepository = expenseRepository;
	}

	@WorkerThread
	public Bill find(final String uuid) {
		return repository.find(billTable, uuid);
	}

	@WorkerThread
	List<Bill> all() {
		return repository.query(billTable, new Where(null).orderBy(Fields.NAME));
	}

	@WorkerThread
	public long greaterUpdatedAt() {
		return repository.greaterUpdatedAt(billTable);
	}

	@WorkerThread
	public List<Bill> unsync() {
		return repository.unsync(billTable);
	}

	@WorkerThread
	public List<Bill> monthly(final DateTime month) {
		final List<Expense> expenses = expenseRepository.monthly(month);
		Where query = new Where(Fields.INIT_DATE).lessThanOrEq(month.getMillis())
				.and(Fields.END_DATE).greaterThanOrEq(month.getMillis());
		List<Bill> bills = repository.query(billTable, query);
		for (int i = 0; i < bills.size(); i++) {
			Bill bill = bills.get(i);
			boolean billAlreadyPaid = false;
			for (Expense expense : expenses) {
				Bill b = repository.find(billTable, expense.getBillUuid());
				if(b != null && b.getUuid().equals(bill.getUuid())) {
					billAlreadyPaid = true;
					break;
				}
			}
			if(billAlreadyPaid) {
				bills.remove(i);
				i--;
			}
		}

		for (Bill bill : bills)
			bill.setMonth(month);

		return bills;
	}

	@WorkerThread
	public ValidationResult save(Bill bill) {
		ValidationResult result = validate(bill);
		if(result.isValid()) {
			if(bill.getId() == 0 && bill.getUuid() == null)
				bill.setUuid(UUID.randomUUID().toString());
			bill.setSync(false);
			repository.saveAtDatabase(billTable, bill);
		}
		return result;
	}

	@NonNull
	private ValidationResult validate(Bill bill) {
		ValidationResult result = new ValidationResult();
		if(StringUtils.isEmpty(bill.getName()))
			result.addError(ValidationError.NAME);
		if(bill.getAmount() <= 0)
			result.addError(ValidationError.AMOUNT);
		if(bill.getDueDate() <= 0)
			result.addError(ValidationError.DUE_DATE);
		if(bill.getInitDate() == null)
			result.addError(ValidationError.INIT_DATE);
		if(bill.getEndDate() == null)
			result.addError(ValidationError.END_DATE);
		if(bill.getInitDate() != null && bill.getEndDate() != null && bill.getInitDate().isAfter(bill.getEndDate()))
			result.addError(ValidationError.INIT_DATE_GREATER_THAN_END_DATE);
		return result;
	}

	@WorkerThread
	@Override
	public ValidationResult syncAndSave(final Bill unsyncBill) {
		ValidationResult result = validate(unsyncBill);
		if(!result.isValid()) {
			Log.Companion.warning("Bill sync validation failed", unsyncBill.getData() + "\nerrors: " + result.getErrorsAsString());
			return result;
		}

		Bill bill = find(unsyncBill.getUuid());
		if(bill != null && bill.getId() != unsyncBill.getId()) {
			if(bill.getUpdatedAt() != unsyncBill.getUpdatedAt())
				Log.Companion.warning("Bill overwritten", unsyncBill.getData());
			unsyncBill.setId(bill.getId());
		}

		unsyncBill.setSync(true);
		repository.saveAtDatabase(billTable, unsyncBill);

		return result;
	}
}

File: app/src/main/java/br/com/jonathanzanella/myexpenses/bill/BillRepository.kt
package br.com.jonathanzanella.myexpenses.bill

import android.support.annotation.WorkerThread
import br.com.jonathanzanella.myexpenses.MyApplication
import br.com.jonathanzanella.myexpenses.database.ModelRepository
import br.com.jonathanzanella.myexpenses.expense.ExpenseRepository
import br.com.jonathanzanella.myexpenses.log.Log
import br.com.jonathanzanella.myexpenses.validations.ValidationError
import br.com.jonathanzanella.myexpenses.validations.ValidationResult
import org.apache.commons.lang3.StringUtils
import org.joda.time.DateTime
import java.util.*

open class BillRepository(private val expenseRepository: ExpenseRepository) : ModelRepository<Bill> {
    @WorkerThread
    fun find(uuid: String): Bill? {
        return MyApplication.database.billDao().find(uuid).blockingFirst()
    }

    @WorkerThread
    fun all(): List<Bill> {
        return MyApplication.database.billDao().all().blockingFirst()
    }

    @WorkerThread
    fun greaterUpdatedAt(): Long {
        return MyApplication.database.billDao().greaterUpdatedAt().blockingFirst().updatedAt
    }

    @WorkerThread
    fun unsync(): List<Bill> {
        return MyApplication.database.billDao().unsync().blockingFirst()
    }

    @WorkerThread
    fun monthly(month: DateTime): List<Bill> {
        val expenses = expenseRepository.monthly(month)
        val bills = MyApplication.database.billDao().monthly(month.millis, month.millis).blockingFirst() as MutableList<Bill>
        var i = 0
        while (i < bills.size) {
            val bill = bills[i]
            val billAlreadyPaid = expenses
                    .map { find(it.billUuid!!) }
                    .any { it != null && it.uuid == bill.uuid }
            if (billAlreadyPaid) {
                bills.removeAt(i)
                i--
            }
            i++
        }

        for (bill in bills)
            bill.month = month

        return bills
    }

    @WorkerThread
    fun save(bill: Bill): ValidationResult {
        val result = validate(bill)
        if (result.isValid) {
            if (bill.id == 0L && bill.uuid == null)
                bill.uuid = UUID.randomUUID().toString()
            bill.sync = false
            bill.id = MyApplication.database.billDao().saveAtDatabase(bill)
        }
        return result
    }

    private fun validate(bill: Bill): ValidationResult {
        val result = ValidationResult()
        if (StringUtils.isEmpty(bill.name))
            result.addError(ValidationError.NAME)
        if (bill.amount <= 0)
            result.addError(ValidationError.AMOUNT)
        if (bill.dueDate <= 0)
            result.addError(ValidationError.DUE_DATE)
        if (bill.initDate == null)
            result.addError(ValidationError.INIT_DATE)
        if (bill.endDate == null)
            result.addError(ValidationError.END_DATE)
        if (bill.initDate != null && bill.endDate != null && bill.initDate!!.isAfter(bill.endDate))
            result.addError(ValidationError.INIT_DATE_GREATER_THAN_END_DATE)
        return result
    }

    @WorkerThread
    override fun syncAndSave(unsync: Bill): ValidationResult {
        val result = validate(unsync)
        if (!result.isValid) {
            Log.warning("Bill sync validation failed", unsync.getData() + "\nerrors: " + result.errorsAsString)
            return result
        }

        val bill = find(unsync.uuid!!)
        if (bill != null && bill.id != unsync.id) {
            if (bill.updatedAt != unsync.updatedAt)
                Log.warning("Bill overwritten", unsync.getData())
            unsync.id = bill.id
        }

        unsync.sync = true
        MyApplication.database.billDao().saveAtDatabase(unsync)

        return result
    }
}